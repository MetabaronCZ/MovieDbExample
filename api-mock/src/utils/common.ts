import { Request } from 'express';

export type QueryParam = Request['query'][number];
export type Query = Record<string, QueryParam>;

const wait = (delay: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const requestDelay = (cb: () => void): void => {
  const delay = 800 + Math.floor(400 * Math.random());
  wait(delay)
    .then(cb)
    .catch((error: unknown) => {
      console.error(error);
    });
};

// extract simple string value from Express query parameter result
export const parseQueryParam = (param: QueryParam): string => {
  const value = Array.isArray(param) ? param[0] : param;
  return 'string' === typeof value ? value : '';
};

type Entries<T extends Record<string, unknown>> = {
  readonly [K in keyof T]: [K, T[K]];
}[keyof T][];

// extract object entries, correctly typed
export const getObjectEntries = <T extends Record<string, unknown>>(
  obj: T,
): Entries<T> => {
  return Object.entries(obj) as Entries<T>;
};
