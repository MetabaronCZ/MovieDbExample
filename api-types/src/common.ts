export const perPages = [25, 50, 100] as const;
export const defaultPerPage = perPages[0];

export const sortDirections = ['ascending', 'descending'] as const;
export type SortDirection = (typeof sortDirections)[number];
