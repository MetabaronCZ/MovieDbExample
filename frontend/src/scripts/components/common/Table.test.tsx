import { fireEvent, getByText, queryByText } from '@testing-library/dom';
import { describe, expect, it, jest } from '@jest/globals';

import { Table } from 'components/common/Table';
import { TestComponentWrapper, waitForComponent } from 'test-utils/component';
import { act } from '@testing-library/react';

interface TableData {
  readonly x: string;
  readonly y: number;
}

describe('components/common/Table', () => {
  it('should render properly', async () => {
    const { container } = await waitForComponent(
      <Table<TableData>
        columns={{
          x: { title: 'Column X', render: (value) => value },
          y: { title: 'Column Y', render: (value) => value },
        }}
        data={[
          { x: 'Value A', y: 112 },
          { x: 'Value B', y: 212 },
          { x: 'Value C', y: 312 },
        ]}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const column1 = getByText(container, 'Column X');
    expect(column1).toBeInTheDocument();
    const column2 = getByText(container, 'Column Y');
    expect(column2).toBeInTheDocument();

    const value1 = getByText(container, 'Value A');
    expect(value1).toBeInTheDocument();
    const value2 = getByText(container, 'Value B');
    expect(value2).toBeInTheDocument();
    const value3 = getByText(container, 'Value C');
    expect(value3).toBeInTheDocument();

    const value4 = getByText(container, '112');
    expect(value4).toBeInTheDocument();
    const value5 = getByText(container, '212');
    expect(value5).toBeInTheDocument();
    const value6 = getByText(container, '312');
    expect(value6).toBeInTheDocument();
  });

  it('should render error state', async () => {
    const { container } = await waitForComponent(
      <Table<TableData>
        error="Some error message!"
        columns={{
          x: { title: 'Column X', render: (value) => value },
          y: { title: 'Column Y', render: (value) => value },
        }}
        data={[{ x: 'Value A', y: 112 }]}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const column1 = getByText(container, 'Column X');
    expect(column1).toBeInTheDocument();
    const column2 = getByText(container, 'Column Y');
    expect(column2).toBeInTheDocument();

    const someValue = queryByText(container, 'Value A');
    expect(someValue).not.toBeInTheDocument();

    const error = getByText(container, 'Some error message!');
    expect(error).toBeInTheDocument();
  });

  it('should render loading state', async () => {
    const { container } = await waitForComponent(
      <Table<TableData>
        isLoading
        columns={{
          x: { title: 'Column X', render: (value) => value },
          y: { title: 'Column Y', render: (value) => value },
        }}
        data={[{ x: 'Value A', y: 112 }]}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const column1 = getByText(container, 'Column X');
    expect(column1).toBeInTheDocument();
    const column2 = getByText(container, 'Column Y');
    expect(column2).toBeInTheDocument();

    const someValue = getByText(container, 'Value A');
    expect(someValue).toBeInTheDocument();

    const content = container.childNodes[0];
    expect(content).toHaveStyle({ opacity: '0.65' });
  });

  it('should render formatted values', async () => {
    const { container } = await waitForComponent(
      <Table<TableData>
        columns={{
          x: { title: 'Column X', render: (value) => `-${value}-` },
          y: { title: 'Column Y', render: (value) => 10 * value },
        }}
        data={[{ x: 'Value A', y: 112 }]}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const value1 = getByText(container, '-Value A-');
    expect(value1).toBeInTheDocument();

    const value2 = getByText(container, '1120');
    expect(value2).toBeInTheDocument();
  });

  it('should sort data', async () => {
    const onSort = jest.fn();

    const { container } = await waitForComponent(
      <Table<TableData>
        columns={{
          x: { title: 'Column X', sort: true, render: (value) => value },
          y: { title: 'Column Y', sort: true, render: (value) => value },
        }}
        data={[
          { x: 'Value A', y: 112 },
          { x: 'Value B', y: 212 },
          { x: 'Value C', y: 312 },
        ]}
        sort="x"
        sortDirection="ascending"
        onSort={onSort}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();
    expect(onSort).not.toBeCalled();

    const header1 = getByText(container, 'Column X');
    expect(header1).toBeInTheDocument();

    const header2 = getByText(container, 'Column Y');
    expect(header2).toBeInTheDocument();

    const button1 = header1.closest('button');
    const button2 = header2.closest('button');

    act(() => {
      if (button1) {
        fireEvent.click(button1);
      }
    });
    expect(onSort).toBeCalledTimes(1);
    expect(onSort).toBeCalledWith('x', 'descending');

    act(() => {
      if (button2) {
        fireEvent.click(button2);
      }
    });
    expect(onSort).toBeCalledTimes(2);
    expect(onSort).toBeCalledWith('y', 'ascending');
  });
});
