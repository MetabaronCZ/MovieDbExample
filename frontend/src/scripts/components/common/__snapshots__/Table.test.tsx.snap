// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/common/Table should render error state 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: 8px;
}

.c1 {
  display: flex;
  flex-direction: row;
  gap: 16px;
}

.c3 {
  display: flex;
  flex-direction: row;
  flex: 1;
  gap: 16px;
}

.c7 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
}

.c6 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  font-weight: 700;
}

.c2 {
  padding-bottom: 8px;
  border-bottom: 1px solid #dae0e3;
}

.c4 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  min-width: 0;
  gap: 8px;
}

.c5 {
  font-weight: 700;
  line-height: 32px;
}

.c8 {
  font-style: italic;
  text-align: center;
}

.c9 {
  color: #f00;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <div
        class="c3 c4 c5"
      >
        <span
          class="c6"
        >
          Column X
        </span>
      </div>
      <div
        class="c3 c4 c5"
      >
        <span
          class="c6"
        >
          Column Y
        </span>
      </div>
    </div>
    <p
      class="c7 c8 c9"
    >
      Some error message!
    </p>
  </div>
</div>
`;

exports[`components/common/Table should render formatted values 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: 8px;
}

.c1 {
  display: flex;
  flex-direction: row;
  gap: 16px;
}

.c3 {
  display: flex;
  flex-direction: row;
  flex: 1;
  gap: 16px;
}

.c7 {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.c6 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  font-weight: 700;
}

.c2 {
  padding-bottom: 8px;
  border-bottom: 1px solid #dae0e3;
}

.c8:hover {
  background-color: #f2f7fa;
}

.c4 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  min-width: 0;
  gap: 8px;
}

.c5 {
  font-weight: 700;
  line-height: 32px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <div
        class="c3 c4 c5"
      >
        <span
          class="c6"
        >
          Column X
        </span>
      </div>
      <div
        class="c3 c4 c5"
      >
        <span
          class="c6"
        >
          Column Y
        </span>
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        class="c1 c8"
      >
        <div
          class="c3 c4"
        >
          -Value A-
        </div>
        <div
          class="c3 c4"
        >
          1120
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/common/Table should render loading state 1`] = `
.c1 {
  display: flex;
  flex-direction: column;
  gap: 8px;
}

.c2 {
  display: flex;
  flex-direction: row;
  gap: 16px;
}

.c4 {
  display: flex;
  flex-direction: row;
  flex: 1;
  gap: 16px;
}

.c8 {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.c0 {
  opacity: 0.65;
  pointer-events: none;
}

.c7 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  font-weight: 700;
}

.c3 {
  padding-bottom: 8px;
  border-bottom: 1px solid #dae0e3;
}

.c9:hover {
  background-color: #f2f7fa;
}

.c5 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  min-width: 0;
  gap: 8px;
}

.c6 {
  font-weight: 700;
  line-height: 32px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2 c3"
      >
        <div
          class="c4 c5 c6"
        >
          <span
            class="c7"
          >
            Column X
          </span>
        </div>
        <div
          class="c4 c5 c6"
        >
          <span
            class="c7"
          >
            Column Y
          </span>
        </div>
      </div>
      <div
        class="c8"
      >
        <div
          class="c2 c9"
        >
          <div
            class="c4 c5"
          >
            Value A
          </div>
          <div
            class="c4 c5"
          >
            112
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/common/Table should render properly 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: 8px;
}

.c1 {
  display: flex;
  flex-direction: row;
  gap: 16px;
}

.c3 {
  display: flex;
  flex-direction: row;
  flex: 1;
  gap: 16px;
}

.c7 {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.c6 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  font-weight: 700;
}

.c2 {
  padding-bottom: 8px;
  border-bottom: 1px solid #dae0e3;
}

.c8:hover {
  background-color: #f2f7fa;
}

.c4 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  min-width: 0;
  gap: 8px;
}

.c5 {
  font-weight: 700;
  line-height: 32px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <div
        class="c3 c4 c5"
      >
        <span
          class="c6"
        >
          Column X
        </span>
      </div>
      <div
        class="c3 c4 c5"
      >
        <span
          class="c6"
        >
          Column Y
        </span>
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        class="c1 c8"
      >
        <div
          class="c3 c4"
        >
          Value A
        </div>
        <div
          class="c3 c4"
        >
          112
        </div>
      </div>
      <div
        class="c1 c8"
      >
        <div
          class="c3 c4"
        >
          Value B
        </div>
        <div
          class="c3 c4"
        >
          212
        </div>
      </div>
      <div
        class="c1 c8"
      >
        <div
          class="c3 c4"
        >
          Value C
        </div>
        <div
          class="c3 c4"
        >
          312
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/common/Table should sort data 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: 8px;
}

.c1 {
  display: flex;
  flex-direction: row;
  gap: 16px;
}

.c3 {
  display: flex;
  flex-direction: row;
  flex: 1;
  gap: 16px;
}

.c9 {
  display: flex;
  flex-direction: column;
  gap: 0px;
}

.c12 {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.c6 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c6:disabled {
  cursor: not-allowed;
}

.c10 {
  position: relative;
  width: 16px;
  height: 8px;
}

.c10>svg {
  position: absolute;
  right: 0;
  height: 16px;
}

.c11 {
  position: relative;
  width: 16px;
  height: 8px;
}

.c11>svg {
  position: absolute;
  right: 0;
  bottom: 0;
  height: 16px;
}

.c7 {
  display: flex;
  align-items: center;
}

.c7:focus {
  outline: 1px dashed #002d56;
}

.c8 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  font-weight: 700;
}

.c2 {
  padding-bottom: 8px;
  border-bottom: 1px solid #dae0e3;
}

.c13:hover {
  background-color: #f2f7fa;
}

.c4 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  min-width: 0;
  gap: 8px;
}

.c5 {
  font-weight: 700;
  line-height: 32px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <div
        class="c3 c4 c5"
      >
        <button
          class="c6 c7"
          title="Seřadit sestupně"
          type="button"
        >
          <span
            class="c8"
          >
            Column X
          </span>
          <div
            class="c9"
          >
            <div
              class="c10"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-sort-up fa-lg "
                color="#002d56"
                data-icon="sort-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M182.6 41.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8l256 0c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="c11"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-sort-down fa-lg "
                color="#98c3d9"
                data-icon="sort-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M182.6 470.6c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-9.2-9.2-11.9-22.9-6.9-34.9s16.6-19.8 29.6-19.8l256 0c12.9 0 24.6 7.8 29.6 19.8s2.2 25.7-6.9 34.9l-128 128z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </button>
      </div>
      <div
        class="c3 c4 c5"
      >
        <button
          class="c6 c7"
          title="Seřadit vzestupně"
          type="button"
        >
          <span
            class="c8"
          >
            Column Y
          </span>
          <div
            class="c9"
          >
            <div
              class="c10"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-sort-up fa-lg "
                color="#98c3d9"
                data-icon="sort-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M182.6 41.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8l256 0c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="c11"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-sort-down fa-lg "
                color="#98c3d9"
                data-icon="sort-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M182.6 470.6c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-9.2-9.2-11.9-22.9-6.9-34.9s16.6-19.8 29.6-19.8l256 0c12.9 0 24.6 7.8 29.6 19.8s2.2 25.7-6.9 34.9l-128 128z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </button>
      </div>
    </div>
    <div
      class="c12"
    >
      <div
        class="c1 c13"
      >
        <div
          class="c3 c4"
        >
          Value A
        </div>
        <div
          class="c3 c4"
        >
          112
        </div>
      </div>
      <div
        class="c1 c13"
      >
        <div
          class="c3 c4"
        >
          Value B
        </div>
        <div
          class="c3 c4"
        >
          212
        </div>
      </div>
      <div
        class="c1 c13"
      >
        <div
          class="c3 c4"
        >
          Value C
        </div>
        <div
          class="c3 c4"
        >
          312
        </div>
      </div>
    </div>
  </div>
</div>
`;
